### PR TITLE
Batched CPU reductions

### LINKED ISSUE #38716: torch.mean returns a wildly incorrect result 0.3277 on YCbCr version of CIFAR10 on CPU with dtype=float32
## 🐛 Bug

When I load CIFAR10 training dataset, convert it to YCbCr, convert it to tensors, transpose it to HWC format and turn it into one large CPU float32 tensor and calculate its channels means, the result is wildly incorrect.

Also, the incorrect result it produces is (0.3277, 0.3277, 0.3277) - the same as reported in https://github.com/pytorch/pytorch/issues/10248, where the author doesn't even convert it to YCbCr. I think this suggests that this is just some bug in `torch.mean`. Also, on my local machine I only encounter this bug if I try to calculate mean for all 3 channels at once. If I instead calculate it for only one channel, I don't get 0.3277.

## To Reproduce

Code snippet:

```python
from typing import Tuple
from PIL.Image import Image
from einops import rearrange # pip install einops
import torch
from torchvision.datasets import CIFAR10
from torchvision.transforms.functional import to_tensor

tv_ds = CIFAR10(".", download=True)
YCbCr_images: Tuple[Image, ...] = tuple(image.convert("YCbCr") for image, label in tv_ds)
tensors: Tuple[torch.Tensor, ...] = tuple(
    rearrange(to_tensor(image), "c h w -> h w c") for image in YCbCr_images
)
x = torch.stack(tensors).unsqueeze(0)  # 1 × 50000 × 32 × 32 × 3                                                                                                                                                                             
x_flattened = x.reshape(-1, 3)
print(x_flattened.mean(dim=0))  # (0.3277, 0.3277, 0.3277)                                                                                                                                                                                   
for channel in range(3):
    print(x_flattened[:, channel].mean())  # 0.3277, 0.3277, 0.3277 on google colab, but 0.4794, 0.4916, 0.5037 on my local machine                                                                                                                                                                     
print(x_flattened.double().mean(dim=0))  # (0.4790, 0.4809, 0.5077)                                                                                                                                                                          
print(x_flattened.cuda().mean(dim=0))  # (0.4790, 0.4809, 0.5077)
```

[Google colab link](https://colab.research.google.com/drive/1hV54M8zH3Cpx6n-A-JHSSuCVnZ1oBvdC?usp=sharing)

## Expected behavior

I expect to get the correct result rather than (0.3277, 0.3277, 0.3277).

## Environment

On my local machine:

```
Collecting environment information...
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Debian GNU/Linux 10 (buster)
GCC version: (Debian 8.3.0-6) 8.3.0
CMake version: version 3.13.4

Python version: 3.8
Is CUDA available: Yes
CUDA runtime version: 9.2.148
GPU models and configuration: 
GPU 0: GeForce RTX 2080 SUPER
GPU 1: GeForce GTX 1070

Nvidia driver version: 418.113
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.16.2
[conda] cudatoolkit               10.1.243             h6bb024c_0  
[conda] mkl                       2020.0                      166    conda-forge
[conda] numpy                     1.18.1           py38h95a1406_0    conda-forge
[conda] pytorch                   1.4.0           py3.8_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] pytorch-ignite            0.3.0                    pypi_0    pypi
[conda] pytorch-lightning         0.7.1                    pypi_0    pypi
[conda] torchvision               0.5.0                py38_cu101    pytorch
```

On google colab:

```
Collecting environment information...
PyTorch version: 1.5.0+cu101
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: Tesla K80
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.18.4
[pip3] torch==1.5.0+cu101
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[pip3] torchvision==0.6.0+cu101
[conda] Could not collect
```

cc @ezyang @gchanan @zou3519 @nairbv

### PR DESCRIPTION
**Pull request by emcastillo**
Fixes #38716 and #37234

In some operations, just doing a running reduction over a large array can cause precision issues leading to incorrect results. This happens in the outer reduction case, as the reduction is parallelized across columns and there are no partial results.

This PR batches these operations so we have several intermediate results that can lead to more stable results and fewer precision issues. The changes are minimal and I believe it is better to do it here than in the specific reduction routine such as `sum`.

TODO: Decide a correct granularity, the 256 right now was put there almost randomly
Measure performance.
Test.

This PR is for discussing this alternative mostly, please give your thoughts and I will work on it.

### CODE REVIEW

File: aten/src/ATen/native/cpu/Reduce.h [hunk 1]
```diff
@@ -36,30 +36,35 @@ static inline void reduction128(char** data, int64_t n, int64_t stride, func_t o
...
-    acc[2] = vop(acc[2], Vec::loadu(ptr + (2 * Vec::size() * sizeof(scalar_t))));
-    acc[3] = vop(acc[3], Vec::loadu(ptr + (3 * Vec::size() * sizeof(scalar_t))));
-  }
-  if (reduce) {
-    scalar_t buffer[Vec::size()];
-    acc[0] = vop(vop(acc[0], acc[1]), vop(acc[2], acc[3]));
-    acc[0].store(buffer);
-    for (int j = 1; j < Vec::size(); j++) {
-      buffer[0] = op(buffer[0], buffer[j]);
+  uint64_t block_size = 256;
```
Comment (emcastillo) [inline aten/src/ATen/native/cpu/Reduce.h:39]: This needs to be discussed

**Comment by emcastillo** [comment]
Hi @peterbell10 ,
I only did the batching for the `outer_reduction` case, other cases with the `basic_loop` reduction will not improve the error.
Let me look it and tune it so you can make a comparison if you can :).

Thanks!

**Comment by emcastillo** [comment]
Yeah, you are right, let me give it a twist

**Comment by emcastillo** [comment]
I added some code that removes the `basic_loop` and just adds a simplified version of `reduction_128` but non vectorized. This also allows to get rid of the code adapting the parameters from the reduction to the Element-wise based loop that `basic_loop` is.

I still have to measure performance.

**Comment by emcastillo** [comment]
seems that some tests are affected, I will try to fix it

File: aten/src/ATen/native/cpu/Reduce.h [hunk 2]
```diff
@@ -267,9 +293,7 @@ void binary_kernel_reduce_vec(TensorIterator& iter, func_t op, vec_func_t vop, d
       vectorized_outer_reduction(data, inner_stride, size0, size1, op, vop);
     } else {
       UNARY_OUTER_LOOP(data, outer_strides, size1, [&] {
-        char* ptrs[3] = { data[0], data[0], data[1] };
-        int64_t inner_strides[3] = { strides[0], strides[0], strides[1] };
-        basic_loop(ptrs, inner_strides, 0, size0, op);
```
Comment (emcastillo) [inline aten/src/ATen/native/cpu/Reduce.h:272]: The errors happen in this reduction replacement

File: aten/src/ATen/native/cpu/Reduce.h [hunk 3]
```diff
@@ -267,9 +293,15 @@ void binary_kernel_reduce_vec(TensorIterator& iter, func_t op, vec_func_t vop, d
...
+        if (strides[0] > 0) {
```
Comment (emcastillo) [inline aten/src/ATen/native/cpu/Reduce.h:296]: I don't feel happy with this change so I am open to suggestions.

**Comment by peterbell10** [comment]
Here is the error comparison from your latest commit.
Errors

**Comment by emcastillo** [comment]
Thanks!
Tunning the `block_size` will have a great impact on the error measurements, I selected 256 without caring about it too much.
This PR is a general solution that might apply to other routines than sum, (although there doesn't seem to be other ones that could be greatly affected by this issue).

**Comment by ngimel** [comment]
#39516 achieving the same goal was merged, can we close this?

**Comment by emcastillo** [comment]
I am fine with closing it, I was just wondering that since this is a generic solution to the reduction algorithm, other reduction routines could benefit from this, but I can't think of other than sum/prod.

**Pull request closed without merging**